src/widget/openglwindow.cpp

```diff
@@ -0,0 +1,75 @@
+#include "widget/openglwindow.h"
+
+#include <QResizeEvent>
+
+#include "widget/tooltipqopengl.h"
+#include "widget/wglwidget.h"
+
+OpenGLWindow::OpenGLWindow(WGLWidget* widget)
+        : m_pWidget(widget) {
+    QSurfaceFormat format;
+    format.setVersion(2, 1);
+    format.setProfile(QSurfaceFormat::CoreProfile);
+    setFormat(format);
+}
+
+OpenGLWindow::~OpenGLWindow() {
+}
+
+void OpenGLWindow::clear() {
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+}
+
+void OpenGLWindow::initializeGL() {
+    if (m_pWidget) {
+        m_pWidget->initializeGL();
+    }
+}
+
+void OpenGLWindow::paintGL() {
+    if (m_pWidget && isExposed()) {
+        m_pWidget->renderGL();
+        m_pWidget->swapBuffers();
+        m_pWidget->renderGL();
```

Comment:
this looks to me like the widget is being rendered twice. Or is something else happening here?

Reply:
I don't remember why I did this (so yeah, a comment would have been nice). I will figure it out and leave that comment.

Reply:
This was to avoid flickering when resizing. I have moved the extra call to renderGL and swapBuffers to resizeGL and added a comment to make this more clear.